### PR TITLE
OOM handling changes

### PR DESCRIPTION
**Pull request by glandium**
As discussed in https://github.com/rust-lang/rust/issues/49668#issuecomment-384893456 and subsequent.

This does have codegen implications. Even without the hooks, and with a handler that ignores the arguments, the compiler doesn't eliminate calling `rust_oom` with the `Layout`. Even if it managed to eliminate that, with the hooks, I don't know if the compiler would be able to figure out it can skip it if the hook is never set.

A couple implementation notes:
- I went with explicit enums rather than bools because it makes it clearer in callers what is being requested.
- I didn't know what `feature` to put the hook setting functions behind. (and surprisingly, the compile went through without any annotation on the functions)
- There's probably some bikeshedding to do on the naming.

Cc: @Simonsapin, @sfackler

### CODE REVIEW

File: src/libstd/alloc.rs [hunk 1]
```diff
@@ -13,15 +13,77 @@
 #![unstable(issue = "32838", feature = "allocator_api")]
 
 #[doc(inline)] #[allow(deprecated)] pub use alloc_crate::alloc::Heap;
-#[doc(inline)] pub use alloc_crate::alloc::{Global, oom};
+#[doc(inline)] pub use alloc_crate::alloc::{Global, Layout, oom};
 #[doc(inline)] pub use alloc_system::System;
 #[doc(inline)] pub use core::alloc::*;
 
+use panicking::HOOK_LOCK;
```
Comment (sfackler) [inline src/libstd/alloc.rs]: Is there any particular reason we're reusing this lock?

File: src/libstd/alloc.rs [hunk 2]
```diff
@@ -13,15 +13,77 @@
...
+}
+
+static mut HOOK: Hook = Hook::Default;
+
+/// Registers a custom OOM hook, replacing any that was previously registered.
+///
+/// The OOM hook is invoked when an infallible memory allocation fails.
+/// The default hook prints a message to standard error and aborts the
+/// execution, but this behavior can be customized with the [`set_hook`]
+/// and [`take_hook`] functions.
```
Comment (sfackler) [inline src/libstd/alloc.rs]: The hook currently isn't allowed to unwind IIRC. Probably worth documenting.

File: src/libstd/alloc.rs [hunk 3]
```diff
@@ -13,15 +13,77 @@
...
+
+        match hook {
+            Hook::Default => Box::new(default_hook),
+            Hook::Custom(ptr) => Box::from_raw(ptr),
+        }
+    }
+}
+
+fn default_hook(_: Layout) -> ! {
+    rtabort!("memory allocation failed")
```
Comment (sfackler) [inline src/libstd/alloc.rs]: Might as well update this to something like `memory allocation of {} bytes failed`.

**Comment by sfackler** [comment]
cc @Gankro for the RawVec changes

File: src/libstd/alloc.rs (hunk 1 above)
Comment (glandium) [reply src/libstd/alloc.rs]: I saw no particular reason *not* to reuse it.

File: src/libstd/alloc.rs (hunk 2 above)
Comment (SimonSapin) [reply src/libstd/alloc.rs]: Isn’t it really? Why?

We have accepted an RFC for `oom=panic` for the use case of handling OOM with `catch_unwind`: https://github.com/rust-lang/rust/issues/48043, https://github.com/rust-lang/rfcs/blob/master/text/2116-alloc-me-maybe.md#oompanic

File: src/libstd/alloc.rs (hunk 1 above)
Comment (sfackler) [reply src/libstd/alloc.rs]: OOM hooks and panic hooks are totally unrelated concepts :P Let's avoid the overlap.

File: src/libstd/alloc.rs (hunk 2 above)
Comment (sfackler) [reply src/libstd/alloc.rs]: It's definitely intended to work, but there's a comment over here claiming we die in MIR transforms if you try to actually do it: https://github.com/rust-lang/rust/blob/master/src/liballoc/alloc.rs#L104-L121

Comment (sfackler) [reply src/libstd/alloc.rs]: I don't actually know why Box still needs a special lang item to allocate though - maybe we can kill off exchange_alloc entirely at this point?

Comment (SimonSapin) [reply src/libstd/alloc.rs]: Maybe `box foo` syntax needs a lang item? It’s how `Box::new` is defined.

Comment (glandium) [reply src/libstd/alloc.rs]: > The hook currently isn't allowed to unwind IIRC. Probably worth documenting.

If the hook unwinding crashes the compiler, but the intent is for it to be allowed per the `oom=panic` RFC, it probably makes sense to leave any comment about unwinding out of the documentation.

**Comment by glandium** [comment]
So interestingly, unwinding *does* work:
```rust
#![feature(allocator_api)]

fn main() {
    std::alloc::set_hook(Box::new(|layout| { panic!("oom {}", layout.size()) }));
    let result = std::panic::catch_unwind(|| {
        let v = Vec::<u8>::with_capacity(100000000000000);
        println!("{:p}", &v[..]);
    });
    println!("{:?}", result);
}
```

On my machine, with rustc built with the patches here, this prints the following:
```
thread 'main' panicked at 'oom 100000000000000', src/main.rs:4:46
note: Run with `RUST_BACKTRACE=1` for a backtrace.
Err(Any)
```

**Comment by sfackler** [comment]
Ah sweet, seems like it may just be an out of date comment then!

**Comment by SimonSapin** [comment]
Did you mean to change the `book` and `llvm` submodules? r=me as of ccf1c1b87b65342322e99c83e0169d5726c9d41f except for this.

**Comment by glandium** [comment]
Dammit, I keep doing this on rebase.

**Comment by Amanieu** [comment]
Have a look at #30801 which previously implemented OOM handler hooks (which then got removed later, somehow).

Two things stand out in particular:
- There were concerns that `rtabort!` might itself allocate memory.
- You should use an atomic instead of a `RwLock`, the latter might be lazily initialized and allocate memory on some platforms.

**Comment by glandium** [comment]
The hooks were removed in #42727 without an explanation...

About `rtabort`, well, whether it allocates or not is not something that needs to be solved here, since this PR doesn't change anything there.

About the `RWLock`, fair enough.

**Comment by glandium** [comment]
Should I restore the function names that were used before? `alloc::set_hook` is probably too generic of a name.

**Comment by Amanieu** [comment]
> About `rtabort`, well, whether it allocates or not is not something that needs to be solved here, since this PR doesn't change anything there.

I specifically used custom implementations ([unix](https://github.com/rust-lang/rust/pull/30801/files#diff-28f2fd684ad47d385427678d96d2dcd4R49) [windows](https://github.com/rust-lang/rust/pull/30801/files#diff-80b6cd62674354b8eb27dab8b55a11e0R47)) instead of `rtabort` to ensure that they did not allocate memory.

**Comment by glandium** [comment]
> I specifically used custom implementations (unix windows) instead of rtabort to ensure that they did not allocate memory.

As I said, I'm not adding the call to rtabort. It's already there. I'm merely moving it. If you feel strongly about this, please file a separate issue.

**Comment by SimonSapin** [comment]
@bors r+

Looks good!

**Pull request merged**